# Welcome to a lab_🐼s!

Last week, you did amazing work in setting up your computer to do Data Science!  Let's now do Data Science on your computer for the first time!

A few tips to remember:

- Refer to your **lecture notebook** and the **pandas cheat sheet** to help you out with the code!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

<hr style="color: #DD3403;">

# Part 0: Your Lab Group!

We find it's best when you're programming with a group of people who you can help out and can help you out!

- A key part of Data Science is being able to communicate about your work.  If you are the best programmer in the world, but no one uses your code, then the code you're writing is not getting run.  By learning strong communication skills, you will be able to explain your code, and the underlying data science concepts, to everyone -- data scientists and non-scientists alike!

- Every lab you will need to complete some parts of the lab with your group.  There will be both "group discussions" and "individual reflections" as part of every lab.

## Create Your Group

If you have not done so already, **RIGHT NOW** find *(ideally)* two other people to make a group of **two, three, or four** people.   *(We find groups of three work the best, but you can be a group of two or four -- no solo groups, and no groups of five!*)

- If you're at a group of five, split into a group of two and a group of three.  You can still help each other as a larger group, but the group questions should be answered in your smaller group of 2-4 people.

Once you have your group **introduce yourselves to each other! :)**

## Record Your Group

Since you know Python, edit the following cell to add strings about your group:

In [16]:
# First, meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Tianyi"
ca1_name = "Jerry"
ca2_name = "Vikram"


# As part of your introduction, learn your lab partner's name, major, and
# learn something new about them that you never knew!
# - Groups of 3 are ideal :)
# - However groups of 2 or 4 are fine too!
partner1_name = "Jiaqi He"
partner1_netid = "jiaqi22"
partner1_major = "Statistics"

partner2_name = "Charlotte "
partner2_netid = "cs112"
partner2_major = "Statistics"

partner3_name = ""
partner3_netid = ""
partner3_major = ""

<hr style="color: #DD3403;">

# Part 1: 🧑‍🤝‍🧑 Hello Dataset!

Last week, we explored the "Party Dataset" that you created.  We will continue to use the Party Dataset all semester long, but we want even more datasets -- and we want you to come up with the questions!

- You will come up with questions to ask the class for the "Hello Dataset"!
- *(This was originally on lab_intro, so you can use your answers from there if you got to this part!  If you did not, work with your group right now!)*

## Hello Dataset Overview

Your lab homework will be to fill out a class-wide "Hello Survey".  Right now, with your group, you will come up some of the questions that should be on this survey *(and you will do analysis on your questions next week!)*.

With your group, come up with four questions below!  Before you get started, a few notes:

- The survey will be **not be anonymous** and the results will be shared with the class to do analysis on, so you should **only ask questions you feel comfortable to share with the class**.
- **Questions that are inappropriate** for a classroom environment will be thrown out and you won't earn lab credit points for any question we throw out.  (Keep everything PG/PG-13.)
- You are going to submit **one set of questions for your whole group**.  You must decide on the questions together.

## Q1: Question with a Binary Response

A question with a **binary response** is a question that has only two answers.  For example:
- *Do you prefer cats or dogs?*
- *Were you born in Illinois?*

With your group (you must all decide on a single question together), come up with a question that you think will have **exactly 50%/50%** split between the two binary outcomes.  *(The question that gets closest to 50%/50% will get extra credit!)*

**What is your binary question?**

Do you consider yourself introverted or extroverted?

**Why do you think this is a good question?  Write at least two complete sentences on why you believe your question will have a 50%/50% split.**

Because it is mostly random and designed by gene. As the class does not select student by personalities, it should have 50%/50% split.

## Q2: Question with a Numeric Response

A question with a **numeric response** requires a number for an answer.  For example:

- *What is your height?*
- *What day were you born?*

With your group (you must all decide on a single question together), come up with a question and a prediction on what you think the most frequent answer will be to your question.

**What is your numeric question?**

How old are you?

**What do you think the most frequent answer will be?**

18

**Why do you think this answer will be the most frequent answer?  Write at least one complete sentences on why you believe that it will be the most frequent answer.**

Because most of us are freshman.

## Q3: Question with a (non-numeric) Textual Response

A question with a (non-numeric) **textual response** requires a word or two for an answer.  For example:

- *What is your favorite sport to play?*
- *What is the location of your favorite trip/vacation?*

With your group (you must all decide on a single question together), come up with a question where at least two members of your group share the same answer.

**What is your textual response question?**

What is your major?

**What is the shared answer between your group?  Do you also share this answer (if not, what's your answer to the question)?**

Statistics.

## Q4: Open Question

One final question -- no restrictions on the format!  Is there a great question your group discussed that you did not get to use one of the first three questions but you want to know the classes' answer?

**What is your final question for the whole class?**

What do you want to learn in this class?

## Submit Your Questions to Your TA

Your lab TAs will be grading your lab notebook for your full responses, but we want to make sure to get the survey with your questions out to everyone ASAP!

- Pick one member of your group as the "question submitter".
- If you're the question submitter, complete this Google Form for your group (requires @illinois.edu login): https://forms.gle/p1HvVZFT1jeSWRKL9


<hr style="color: #DD3403;">

# Part 2: Importing the Course Catalog Dataset

In lecture, you learned we will use the `pandas` **library** for working with data.  Libraries are collections of functions and other objects that extend what Python can do by itself.  The `pandas` library is the most widely used library for data science.

To use the `pandas` library, we must **import** it into your notebook.  In literally every lab, you will need to import the library just like we did in lecture -- import pandas as `pd` in the following cell:

In [2]:
# Write the code to import pandas as pd:
import pandas as pd

In lecture, you saw the "Course Catalog" dataset.  Using the `read_csv` function, load the course catalog dataset.  The URL for this dataset is:
```
https://waf.cs.illinois.edu/discovery/course-catalog.csv
```


In [3]:
# Load the course catalog dataset into a variable named `df`:
df = pd.read_csv('https://waf.cs.illinois.edu/discovery/course-catalog.csv')

In [19]:
# Print out the contents of `df`:
df


Year  Term YearTerm Subject  Number                          Name  \
0   2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
1   2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
2   2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
3   2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
4   2022  Fall  2022-fa     AAS     100  Intro Asian American Studies   
..   ...   ...      ...     ...     ...                           ...   
95  2022  Fall  2022-fa    ACCY     201  Accounting and Accountancy I   
96  2022  Fall  2022-fa    ACCY     201  Accounting and Accountancy I   
97  2022  Fall  2022-fa    ACCY     201  Accounting and Accountancy I   
98  2022  Fall  2022-fa    ACCY     201  Accounting and Accountancy I   
99  2022  Fall  2022-fa    ACCY     201  Accounting and Accountancy I   

                                          Description Credit Hours  \
0   Interdisciplinary introduction to the basic co...     3 hours.   
1   Interdisciplinary introduction to the basic co...     3 hours.   
2   Interdisciplinary introduction to the basic co...     3 hours.   
3   Interdisciplinary introduction to the basic co...     3 hours.   
4   Interdisciplinary introduction to the basic co...     3 hours.   
..                                                ...          ...   
95  Develops a foundation for understanding and an...     3 hours.   
96  Develops a foundation for understanding and an...     3 hours.   
97  Develops a foundation for understanding and an...     3 hours.   
98  Develops a foundation for understanding and an...     3 hours.   
99  Develops a foundation for understanding and an...     3 hours.   

                                         Section Info  \
0                                                 NaN   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   
..                                                ...   
95  Credit is not given for both ACCY 201 and ACCY...   
96  Credit is not given for both ACCY 201 and ACCY...   
97  Credit is not given for both ACCY 201 and ACCY...   
98  Credit is not given for both ACCY 201 and ACCY...   
99  Credit is not given for both ACCY 201 and ACCY...   

                                    Degree Attributes  ... Section Status  \
0   Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
1   Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
2   Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
3   Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
4   Social & Beh Sci - Soc Sci, and Cultural Studi...  ...              A   
..                                                ...  ...            ...   
95                                                NaN  ...              A   
96                                                NaN  ...              A   
97                                                NaN  ...              A   
98                                                NaN  ...              A   
99                                                NaN  ...              A   

    Enrollment Status                   Type Type Code Start Time  End Time  \
0                Open  Discussion/Recitation       DIS   09:00 AM  09:50 AM   
1              Closed  Discussion/Recitation       DIS   10:00 AM  10:50 AM   
2                Open  Discussion/Recitation       DIS   10:00 AM  10:50 AM   
3              Closed  Discussion/Recitation       DIS   11:00 AM  11:50 AM   
4                Open  Discussion/Recitation       DIS   11:00 AM  11:50 AM   
..                ...                    ...       ...        ...       ...   
95             Closed  Discussion/Recitation       DIS   08:00 AM  08:50 AM   
96               Open  Discussion/Recitation

## 🔬 Test Case Checkpoint 🔬

Throughout the lab, we will have "checkpoints" at key points.  Most checkpoints will include **test cases** that verify that your notebook appears to be correct.

- The test cases are **NOT** a guarantee that everything is perfect, but we try our very best to catch any common mistakes you might make!
- If you see `AssertionError` error message when running the test case, we provide a helpful hint on what may not be correct.
- If you see a party popper emoji and a message that all tests pass, you passed the checkpoint!
- **Checkpoints** are a good time to check with your group to make everyone gets to the checkpoint together!

In [21]:
## == CHECKPOINT TEST CASES for Part 1 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "Did you define a variable called df?"
assert( len(df) == 12643 ), "It looks like you did not import the correct dataset?"


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


<hr style="color: #DD3403;">

# Part 3: STAT / CS / IS Majors

Data Science DISCOVERY is a unique course that is cross-listed across three departments: Statistics, Computer Science, and Information Sciences.  Let's explore courses from all three of those subjects!

### Puzzle 3.1

Select only the rows that have a `Subject` of `STAT`.  Assign this to the variable name `df_STAT` instead of the usual `df`.
- Make sure your result contains only statistics courses!

In [25]:
df_STAT = df[df["Subject"] == "STAT"]
df_STAT

Year  Term YearTerm Subject  Number                           Name  \
11890  2022  Fall  2022-fa    STAT     100                     Statistics   
11891  2022  Fall  2022-fa    STAT     100                     Statistics   
11892  2022  Fall  2022-fa    STAT     107         Data Science Discovery   
11893  2022  Fall  2022-fa    STAT     107         Data Science Discovery   
11894  2022  Fall  2022-fa    STAT     107         Data Science Discovery   
...     ...   ...      ...     ...     ...                            ...   
12025  2022  Fall  2022-fa    STAT     578           Topics in Statistics   
12026  2022  Fall  2022-fa    STAT     590  Individual Study and Research   
12027  2022  Fall  2022-fa    STAT     593                STAT Internship   
12028  2022  Fall  2022-fa    STAT     593                STAT Internship   
12029  2022  Fall  2022-fa    STAT     599                Thesis Research   

                                             Description    Credit Hours  \
11890  First course in probability and statistics at ...        3 hours.   
11891  First course in probability and statistics at ...        3 hours.   
11892  Data Science Discovery is the intersection of ...        4 hours.   
11893  Data Science Discovery is the intersection of ...        4 hours.   
11894  Data Science Discovery is the intersection of ...        4 hours.   
...                                                  ...             ...   
12025  May be repeated if topics vary. Prerequisite: ...        4 hours.   
12026  Directed reading and research. Approved for le...   0 TO 8 hours.   
12027  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12028  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12029  Approved for S/U grading only. May be repeated...  0 TO 16 hours.   

                                            Section Info  \
11890  Credit is not given for both STAT 100 and any ...   
11891  Credit is not given for both STAT 100 and any ...   
11892                         Same as CS 107 and IS 107.   
11893                         Same as CS 107 and IS 107.   
11894                         Same as CS 107 and IS 107.   
...                                                  ...   
12025  May be repeated if topics vary. Prerequisite: ...   
12026  Approved for letter and S/U grading. May be re...   
12027  Approved for letter and S/U grading. Prerequis...   
12028  Approved for letter and S/U grading. Prerequis...   
12029  Approved for S/U grading only. May be repeated...   

                      Degree Attributes  ... Section Status  \
11890  Quantitative Reasoning I course.  ...              A   
11891  Quantitative Reasoning I course.  ...              A   
11892  Quantitative Reasoning I course.  ...              A   
11893  Quantitative Reasoning I course.  ...              A   
11894  Quantitative Reasoning I course.  ...              A   
...                                 ...  ...            ...   
12025                               NaN  ...              A   
12026                               NaN  ...              A   
12027                               NaN  ...              A   
12028                               NaN  ...              A   
12029                               NaN  ...              A   

       Enrollment Status                   Type Type Code Start Time  \
11890             Closed     Lecture-Discussion       LCD   12:00 PM   
11891             Closed                 Online       ONL   ARRANGED   
11892      CrossListOpen                Lecture       LEC   11:00 AM   
11893             Closed  Laboratory-Discussion       LBD   12:30 PM   
11894             Closed  Laboratory-Discussion       LBD   12:30 PM   
...                  ...                    ...       ...        ...   
12025  Open (Restricted)     Lecture-Discussion        S0   12:00 PM   
12026  Open (Restricted)      Independent Study       IND   ARRANGED   
12027  Open (Restricted)             Intern

### Puzzle 3.2

Select only the rows that have a `Subject` of `CS`.  Assign this to the variable name `df_CS` instead of the usual `df`.
- Make sure your result contains only computer science courses!

In [28]:
df_CS = df[df["Subject"] == "CS"]
df_CS

Year  Term YearTerm Subject  Number                              Name  \
3349  2022  Fall  2022-fa      CS     100      Computer Science Orientation   
3350  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3351  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3352  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
3353  2022  Fall  2022-fa      CS     101  Intro Computing: Engrg &amp; Sci   
...    ...   ...      ...     ...     ...                               ...   
3731  2022  Fall  2022-fa      CS     598                    Special Topics   
3732  2022  Fall  2022-fa      CS     598                    Special Topics   
3733  2022  Fall  2022-fa      CS     598                    Special Topics   
3734  2022  Fall  2022-fa      CS     598                    Special Topics   
3735  2022  Fall  2022-fa      CS     599                   Thesis Research   

                                            Description    Credit Hours  \
3349  Introduction to Computer Science as a field an...        1 hours.   
3350  Fundamental principles, concepts, and methods ...        3 hours.   
3351  Fundamental principles, concepts, and methods ...        3 hours.   
3352  Fundamental principles, concepts, and methods ...        3 hours.   
3353  Fundamental principles, concepts, and methods ...        3 hours.   
...                                                 ...             ...   
3731  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3732  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3733  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3734  Subject offerings of new and developing areas ...   2 TO 4 hours.   
3735    Approved for S/U grading only. May be repeated.  0 TO 16 hours.   

                                           Section Info  \
3349                                                NaN   
3350  Prerequisite: One of MATH 220 or MATH 221 or M...   
3351  Prerequisite: One of MATH 220 or MATH 221 or M...   
3352  Prerequisite: One of MATH 220 or MATH 221 or M...   
3353  Prerequisite: One of MATH 220 or MATH 221 or M...   
...                                                 ...   
3731  May be repeated in the same or separate terms ...   
3732  May be repeated in the same or separate terms ...   
3733  May be repeated in the same or separate terms ...   
3734  May be repeated in the same or separate terms ...   
3735    Approved for S/U grading only. May be repeated.   

                      Degree Attributes  ... Section Status  \
3349                                NaN  ...              A   
3350  Quantitative Reasoning II course.  ...              A   
3351  Quantitative Reasoning II course.  ...              A   
3352  Quantitative Reasoning II course.  ...              A   
3353  Quantitative Reasoning II course.  ...              P   
...                                 ...  ...            ...   
3731                                NaN  ...              A   
3732                                NaN  ...              A   
3733                                NaN  ...              A   
3734                                NaN  ...              A   
3735                                NaN  ...              A   

      Enrollment Status                   Type Type Code Start Time  End Time  \
3349  Open (Restricted)                Lecture       LEC   03:00 PM  03:50 PM   
3350  Open (Restricted)     Lecture-Discussion       LCD   10:00 AM  10:50 AM   
3351               Open  Laboratory-Discussion       LBD   01:00 PM  02:50 PM   
3352               Open  Laboratory-Discussion       LBD   03:00 PM  04:50 PM   
3353             Closed  Laboratory-Discussion       LBD   05:00 PM  06:50 PM   
...                 ...                    ...       ...        ...       ...   
3731             Closed     Lecture-Discussion       LCD   11:00 AM  12:15 PM   
3732  Open (Restricted)     Lecture-Discus

### Puzzle 3.3

Select only the rows that have a `Subject` of `IS`.  Assign this to the variable name `df_IS` instead of the usual `df`.
- Make sure your result contains only information science courses!

In [30]:
df_IS = df[df["Subject"] == "IS"]
df_IS

Year  Term YearTerm Subject  Number  \
6663  2022  Fall  2022-fa      IS     100   
6664  2022  Fall  2022-fa      IS     100   
6665  2022  Fall  2022-fa      IS     100   
6666  2022  Fall  2022-fa      IS     100   
6667  2022  Fall  2022-fa      IS     100   
...    ...   ...      ...     ...     ...   
6907  2022  Fall  2022-fa      IS     596   
6908  2022  Fall  2022-fa      IS     597   
6909  2022  Fall  2022-fa      IS     597   
6910  2022  Fall  2022-fa      IS     597   
6911  2022  Fall  2022-fa      IS     599   

                                                   Name  \
6663   Exploring the iSchool with a Human-Centered Lens   
6664   Exploring the iSchool with a Human-Centered Lens   
6665   Exploring the iSchool with a Human-Centered Lens   
6666   Exploring the iSchool with a Human-Centered Lens   
6667   Exploring the iSchool with a Human-Centered Lens   
...                                                 ...   
6907  Advanced Topics in Human-Centered Design &amp;...   
6908  Advanced Topics in Data Analytics &amp; Data S...   
6909  Advanced Topics in Data Analytics &amp; Data S...   
6910  Advanced Topics in Data Analytics &amp; Data S...   
6911                                    Thesis Research   

                                            Description    Credit Hours  \
6663  This course introduces students to the School ...        1 hours.   
6664  This course introduces students to the School ...        1 hours.   
6665  This course introduces students to the School ...        1 hours.   
6666  This course introduces students to the School ...        1 hours.   
6667  This course introduces students to the School ...        1 hours.   
...                                                 ...             ...   
6907  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6908  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6909  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6910  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6911  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info Degree Attributes  \
6663  Prerequisite: Restricted to Majors Only; First...               NaN   
6664  Prerequisite: Restricted to Majors Only; First...               NaN   
6665  Prerequisite: Restricted to Majors Only; First...               NaN   
6666  Prerequisite: Restricted to Majors Only; First...               NaN   
6667  Prerequisite: Restricted to Majors Only; First...               NaN   
...                                                 ...               ...   
6907  2 or 4 graduate hours. No professional credit....               NaN   
6908  2 or 4 graduate hours. No professional credit....               NaN   
6909  2 or 4 graduate hours. No professional credit....               NaN   
6910  2 or 4 graduate hours. No professional credit....               NaN   
6911  0 to 16 graduate hours. No professional credit...               NaN   

      ... Section Status           Enrollment Status                Type  \
6663  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6664  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6665  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6666  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6667  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
...   ...            ...                         ...                 ...   
6907  ...              A           Open (Restricted)              Online   
6908  ...              A           Open (Restricted)  Lecture-Discussion   
6909  ...              A           Open (Restricted)  Lecture-Discussion   
6910  ...              A           Open (Restricted)              Online   
6911  ...              A           Open (Restricted)   Independent Study   

     Type Code St

### 🔬 Test Case Checkpoint 🔬


In [ ]:
## == CHECKPOINT TEST CASES for Part 3 ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df_CS' in vars() ), "Make certain to name the CS courses df_CS."
assert( 'df_STAT' in vars() ), "Make certain to name the STAT courses df_STAT."
assert( 'df_IS' in vars() ), "Make certain to name the IS courses df_IS."
assert( len(df_STAT[df_STAT.Subject != "STAT"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_CS[df_CS.Subject != "CS"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."
assert( len(df_IS[df_IS.Subject != "IS"] ) == 0 ), "It looks like you did not subset df_STAT to only STAT courses."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

### Puzzle 3.4

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject offers the **fewest** course sections?  The **most** course sections?


CS offers the most course sections and STATS offers the fewest course sections.

### Puzzle 3.5

Looking at the output of the DataFrames for STAT, CS, and IS above, which subject has the highest numbered course?  Is there something similar about all of the highest numbered courses?  Write about what you observed!


All of them offers the same highest numbered course of 599. They are all about thesis research.

<hr style="color: #DD3403;">

# Part 4: Your Major

In this part, you will be exploring **your major** using the Course Catalog dataset.

- Use the primary subject of your **current major** (or, if you're undeclared, your intended major) for this section.
- Your answers will be different than the rest of your group!



### Puzzle 4.1:

Select only the rows that have a `Subject` of your current/intended major.  Assign this to the variable name `mydf` instead of `df`:

In [31]:
mydf = df[df["Subject"] == "STAT"]
mydf

Year  Term YearTerm Subject  Number                           Name  \
11890  2022  Fall  2022-fa    STAT     100                     Statistics   
11891  2022  Fall  2022-fa    STAT     100                     Statistics   
11892  2022  Fall  2022-fa    STAT     107         Data Science Discovery   
11893  2022  Fall  2022-fa    STAT     107         Data Science Discovery   
11894  2022  Fall  2022-fa    STAT     107         Data Science Discovery   
...     ...   ...      ...     ...     ...                            ...   
12025  2022  Fall  2022-fa    STAT     578           Topics in Statistics   
12026  2022  Fall  2022-fa    STAT     590  Individual Study and Research   
12027  2022  Fall  2022-fa    STAT     593                STAT Internship   
12028  2022  Fall  2022-fa    STAT     593                STAT Internship   
12029  2022  Fall  2022-fa    STAT     599                Thesis Research   

                                             Description    Credit Hours  \
11890  First course in probability and statistics at ...        3 hours.   
11891  First course in probability and statistics at ...        3 hours.   
11892  Data Science Discovery is the intersection of ...        4 hours.   
11893  Data Science Discovery is the intersection of ...        4 hours.   
11894  Data Science Discovery is the intersection of ...        4 hours.   
...                                                  ...             ...   
12025  May be repeated if topics vary. Prerequisite: ...        4 hours.   
12026  Directed reading and research. Approved for le...   0 TO 8 hours.   
12027  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12028  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12029  Approved for S/U grading only. May be repeated...  0 TO 16 hours.   

                                            Section Info  \
11890  Credit is not given for both STAT 100 and any ...   
11891  Credit is not given for both STAT 100 and any ...   
11892                         Same as CS 107 and IS 107.   
11893                         Same as CS 107 and IS 107.   
11894                         Same as CS 107 and IS 107.   
...                                                  ...   
12025  May be repeated if topics vary. Prerequisite: ...   
12026  Approved for letter and S/U grading. May be re...   
12027  Approved for letter and S/U grading. Prerequis...   
12028  Approved for letter and S/U grading. Prerequis...   
12029  Approved for S/U grading only. May be repeated...   

                      Degree Attributes  ... Section Status  \
11890  Quantitative Reasoning I course.  ...              A   
11891  Quantitative Reasoning I course.  ...              A   
11892  Quantitative Reasoning I course.  ...              A   
11893  Quantitative Reasoning I course.  ...              A   
11894  Quantitative Reasoning I course.  ...              A   
...                                 ...  ...            ...   
12025                               NaN  ...              A   
12026                               NaN  ...              A   
12027                               NaN  ...              A   
12028                               NaN  ...              A   
12029                               NaN  ...              A   

       Enrollment Status                   Type Type Code Start Time  \
11890             Closed     Lecture-Discussion       LCD   12:00 PM   
11891             Closed                 Online       ONL   ARRANGED   
11892      CrossListOpen                Lecture       LEC   11:00 AM   
11893             Closed  Laboratory-Discussion       LBD   12:30 PM   
11894             Closed  Laboratory-Discussion       LBD   12:30 PM   
...                  ...                    ...       ...        ...   
12025  Open (Restricted)     Lecture-Discussion        S0   12:00 PM   
12026  Open (Restricted)      Independent Study       IND   ARRANGED   
12027  Open (Restricted)             Intern

### Puzzle 4.2:

Select **five random courses** from your major, using `mydf` and pandas functions:

In [32]:
mydf.sample(5)

Year  Term YearTerm Subject  Number                          Name  \
11995  2022  Fall  2022-fa    STAT     431     Applied Bayesian Analysis   
12007  2022  Fall  2022-fa    STAT     443       Professional Statistics   
11987  2022  Fall  2022-fa    STAT     429          Time Series Analysis   
11991  2022  Fall  2022-fa    STAT     430  Topics in Applied Statistics   
11989  2022  Fall  2022-fa    STAT     430  Topics in Applied Statistics   

                                             Description   Credit Hours  \
11995  Introduction to the concepts and methodology o...  3 OR 4 hours.   
12007  This project-based course emphasizes written, ...  3 OR 4 hours.   
11987  Studies theory and data analysis for time seri...  3 OR 4 hours.   
11991  Formulation and analysis of mathematical model...  3 OR 4 hours.   
11989  Formulation and analysis of mathematical model...  3 OR 4 hours.   

                                            Section Info Degree Attributes  \
11995  Same as ASRM 453. 3 undergraduate hours. 4 gra...               NaN   
12007  3 undergraduate hours. 4 graduate hours. Prere...               NaN   
11987  3 undergraduate hours. 4 graduate hours. Prere...               NaN   
11991  3 undergraduate hours. 4 graduate hours. May b...               NaN   
11989  3 undergraduate hours. 4 graduate hours. May b...               NaN   

       ... Section Status           Enrollment Status                   Type  \
11995  ...              A  CrossListOpen (Restricted)     Lecture-Discussion   
12007  ...              A  CrossListOpen (Restricted)  Laboratory-Discussion   
11987  ...              A                      Closed                 Online   
11991  ...              A  CrossListOpen (Restricted)                 Online   
11989  ...              A  CrossListOpen (Restricted)     Lecture-Discussion   

      Type Code Start Time  End Time Days of Week  Room  \
11995       LCD   11:00 AM  12:20 PM           TR   217   
12007       LBD   03:00 PM  03:50 PM          MWF  3217   
11987       ONL   ARRANGED       NaN          NaN   NaN   
11991       ONL   ARRANGED       NaN          NaN   NaN   
11989       LCD   09:00 AM  09:50 AM          MWF  4039   

                            Building    Instructors  
11995               Noyes Laboratory        Park, T  
12007             Everitt Laboratory  Glosemeyer, D  
11987                            NaN         Lee, H  
11991                            NaN    Balamuta, J  
11989  Campus Instructional Facility        Zhao, S  

[5 rows x 27 columns]

### Puzzle 4.3:

Select **the seven lowest numbered courses** from your major:

In [43]:
mydf.nsmallest(7,'Number')

Year  Term YearTerm Subject  Number                    Name  \
11890  2022  Fall  2022-fa    STAT     100              Statistics   
11891  2022  Fall  2022-fa    STAT     100              Statistics   
11892  2022  Fall  2022-fa    STAT     107  Data Science Discovery   
11893  2022  Fall  2022-fa    STAT     107  Data Science Discovery   
11894  2022  Fall  2022-fa    STAT     107  Data Science Discovery   
11895  2022  Fall  2022-fa    STAT     107  Data Science Discovery   
11896  2022  Fall  2022-fa    STAT     107  Data Science Discovery   

                                             Description Credit Hours  \
11890  First course in probability and statistics at ...     3 hours.   
11891  First course in probability and statistics at ...     3 hours.   
11892  Data Science Discovery is the intersection of ...     4 hours.   
11893  Data Science Discovery is the intersection of ...     4 hours.   
11894  Data Science Discovery is the intersection of ...     4 hours.   
11895  Data Science Discovery is the intersection of ...     4 hours.   
11896  Data Science Discovery is the intersection of ...     4 hours.   

                                            Section Info  \
11890  Credit is not given for both STAT 100 and any ...   
11891  Credit is not given for both STAT 100 and any ...   
11892                         Same as CS 107 and IS 107.   
11893                         Same as CS 107 and IS 107.   
11894                         Same as CS 107 and IS 107.   
11895                         Same as CS 107 and IS 107.   
11896                         Same as CS 107 and IS 107.   

                      Degree Attributes  ... Section Status  \
11890  Quantitative Reasoning I course.  ...              A   
11891  Quantitative Reasoning I course.  ...              A   
11892  Quantitative Reasoning I course.  ...              A   
11893  Quantitative Reasoning I course.  ...              A   
11894  Quantitative Reasoning I course.  ...              A   
11895  Quantitative Reasoning I course.  ...              A   
11896  Quantitative Reasoning I course.  ...              A   

       Enrollment Status                   Type Type Code Start Time  \
11890             Closed     Lecture-Discussion       LCD   12:00 PM   
11891             Closed                 Online       ONL   ARRANGED   
11892      CrossListOpen                Lecture       LEC   11:00 AM   
11893             Closed  Laboratory-Discussion       LBD   12:30 PM   
11894             Closed  Laboratory-Discussion       LBD   12:30 PM   
11895             Closed  Laboratory-Discussion       LBD   02:00 PM   
11896             Closed  Laboratory-Discussion       LBD   02:00 PM   

       End Time Days of Week   Room            Building  \
11890  12:50 PM          MWF  THEAT        Lincoln Hall   
11891       NaN          NaN    NaN                 NaN   
11892  11:50 AM          MWF  THEAT        Lincoln Hall   
11893  01:50 PM            W    209           Huff Hall   
11894  01:50 PM            W   2101  Everitt Laboratory   
11895  03:20 PM            W    209           Huff Hall   
11896  03:20 PM            W   2101  Everitt Laboratory   

                             Instructors  
11890                        Flanagan, K  
11891                        Flanagan, K  
11892  Fagen-Ulmschneider, W;Flanagan, K  
11893                            Xing, Y  
11894                             Kiv, D  
11895                             Gao, H  
11896                             Kiv, D  

[7 rows x 27 columns]

### Puzzle 4.4:

Select **the nine highest numbered courses** from your major:

In [44]:
mydf.nlargest(9,'Number')

Year  Term YearTerm Subject  Number                           Name  \
12029  2022  Fall  2022-fa    STAT     599                Thesis Research   
12027  2022  Fall  2022-fa    STAT     593                STAT Internship   
12028  2022  Fall  2022-fa    STAT     593                STAT Internship   
12026  2022  Fall  2022-fa    STAT     590  Individual Study and Research   
12024  2022  Fall  2022-fa    STAT     578           Topics in Statistics   
12025  2022  Fall  2022-fa    STAT     578           Topics in Statistics   
12023  2022  Fall  2022-fa    STAT     571          Multivariate Analysis   
12022  2022  Fall  2022-fa    STAT     558     Risk Modeling and Analysis   
12021  2022  Fall  2022-fa    STAT     555   Applied Stochastic Processes   

                                             Description    Credit Hours  \
12029  Approved for S/U grading only. May be repeated...  0 TO 16 hours.   
12027  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12028  Supervised, off-campus experience in a field i...   0 TO 8 hours.   
12026  Directed reading and research. Approved for le...   0 TO 8 hours.   
12024  May be repeated if topics vary. Prerequisite: ...        4 hours.   
12025  May be repeated if topics vary. Prerequisite: ...        4 hours.   
12023  Inference in multivariate statistical populati...        4 hours.   
12022                    Same as MATH 563. See MATH 563.        4 hours.   
12021                    Same as MATH 564. See MATH 564.        4 hours.   

                                            Section Info Degree Attributes  \
12029  Approved for S/U grading only. May be repeated...               NaN   
12027  Approved for letter and S/U grading. Prerequis...               NaN   
12028  Approved for letter and S/U grading. Prerequis...               NaN   
12026  Approved for letter and S/U grading. May be re...               NaN   
12024  May be repeated if topics vary. Prerequisite: ...               NaN   
12025  May be repeated if topics vary. Prerequisite: ...               NaN   
12023  Prerequisite: STAT 410 and MATH 415, or consen...               NaN   
12022                    Same as MATH 563. See MATH 563.               NaN   
12021                    Same as MATH 564. See MATH 564.               NaN   

       ... Section Status  Enrollment Status                Type Type Code  \
12029  ...              A  Open (Restricted)   Independent Study       IND   
12027  ...              A  Open (Restricted)          Internship       INT   
12028  ...              A  Open (Restricted)          Internship       INT   
12026  ...              A  Open (Restricted)   Independent Study       IND   
12024  ...              A  Open (Restricted)  Lecture-Discussion       LCD   
12025  ...              A  Open (Restricted)  Lecture-Discussion        S0   
12023  ...              A  Open (Restricted)  Lecture-Discussion       LCD   
12022  ...              A      CrossListOpen  Lecture-Discussion       LCD   
12021  ...              A             Closed  Lecture-Discussion       LCD   

      Start Time  End Time Days of Week   Room          Building  \
12029   ARRANGED       NaN          NaN    NaN               NaN   
12027   ARRANGED       NaN          NaN    NaN               NaN   
12028   ARRANGED       NaN          NaN    NaN               NaN   
12026   ARRANGED       NaN          NaN    NaN               NaN   
12024   03:00 PM  03:50 PM          MWF   1090      Lincoln Hall   
12025   12:00 PM  12:50 PM          MWF  106B8  Engineering Hall   
12023   03:00 PM  03:50 PM          MWF  106B8  Engineering Hall   
12022   04:30 PM  05:50 PM           MW    447      Altgeld Hall   
12021   09:30 AM  10:50 AM           TR    147      Altgeld Hall   

                                             Instructors  
12029                                                NaN  
12027  Chronopoulou, A;Glosemeyer, D;Lee, H;Shao, X;U...  
12028  Chronopoulou, A;Glosemeyer, D;Lee, H;Shao, X;U..

### Puzzle 4.5:

Select **a random sample of 10% of the courses** from your major:

In [37]:
mydf.sample(frac=0.1)

Year  Term YearTerm Subject  Number  \
11925  2022  Fall  2022-fa    STAT     361   
11986  2022  Fall  2022-fa    STAT     429   
11955  2022  Fall  2022-fa    STAT     400   
12001  2022  Fall  2022-fa    STAT     433   
11914  2022  Fall  2022-fa    STAT     200   
11945  2022  Fall  2022-fa    STAT     400   
11989  2022  Fall  2022-fa    STAT     430   
12020  2022  Fall  2022-fa    STAT     553   
12010  2022  Fall  2022-fa    STAT     447   
11939  2022  Fall  2022-fa    STAT     400   
11956  2022  Fall  2022-fa    STAT     400   
11910  2022  Fall  2022-fa    STAT     107   
11928  2022  Fall  2022-fa    STAT     361   
11983  2022  Fall  2022-fa    STAT     426   

                                                    Name  \
11925  Probability &amp; Statistics for Computer Science   
11986                               Time Series Analysis   
11955                       Statistics and Probability I   
12001                               Stochastic Processes   
11914                               Statistical Analysis   
11945                       Statistics and Probability I   
11989                       Topics in Applied Statistics   
12020                          Probability and Measure I   
12010                   Data Science Programming Methods   
11939                       Statistics and Probability I   
11956                       Statistics and Probability I   
11910                             Data Science Discovery   
11928  Probability &amp; Statistics for Computer Science   
11983                            Statistical Modeling II   

                                             Description   Credit Hours  \
11925                        Same as CS 361. See CS 361.       3 hours.   
11986  Studies theory and data analysis for time seri...  3 OR 4 hours.   
11955  Introduction to mathematical statistics that d...       4 hours.   
12001  A stochastic process is a random process that ...  3 OR 4 hours.   
11914  Survey of statistical concepts, data analysis,...       3 hours.   
11945  Introduction to mathematical statistics that d...       4 hours.   
11989  Formulation and analysis of mathematical model...  3 OR 4 hours.   
12020  Measures and probabilities; integration and ex...       4 hours.   
12010  The field of data science is revolutionizing s...  3 OR 4 hours.   
11939  Introduction to mathematical statistics that d...       4 hours.   
11956  Introduction to mathematical statistics that d...       4 hours.   
11910  Data Science Discovery is the intersection of ...       4 hours.   
11928                        Same as CS 361. See CS 361.       3 hours.   
11983  This is a continuation in the study of advance...  3 OR 4 hours.   

                                            Section Info  \
11925                        Same as CS 361. See CS 361.   
11986  3 undergraduate hours. 4 graduate hours. Prere...   
11955  Same as MATH 463. 4 undergraduate hours. 4 gra...   
12001  3 undergraduate hours. 4 graduate hours. Prere...   
11914  Credit is not given for both STAT 200 and STAT...   
11945  Same as MATH 463. 4 undergraduate hours. 4 gra...   
11989  3 undergraduate hours. 4 graduate hours. May b...   
12020   Prerequisite: MATH 447 or consent of instructor.   
12010           3 undergraduate hours. 4 graduate hours.   
11939  Same as MATH 463. 4 undergraduate hours. 4 gra...   
11956  Same as MATH 463. 4 undergraduate hours. 4 gra...   
11910                         Same as CS 107 and IS 107.   
11928                        Same as CS 361. See CS 361.   
11983  3 undergraduate hours. 4 graduate hours. Prere...   

                      Degree Attributes  ... Section Status  \
11925                               NaN  ...              A   
11986                               NaN  ...              A   
11955                               NaN  ...              A   
12001                               NaN  ...              A   
11914  Quantitative Reasoning I course.  ...              A   
11945     

### ✔️ Checkpoint ✔️

There is no test cases for Part 4 (since everyone has different majors), but go double check your work!  There's just one section left to go! :)

<hr style="color: #DD3403;">

# Part 5: Growth of the IS major

The IS major is a new relatively new major here at UIUC.  The final puzzle is for you to analyze how much the IS major has grown since 2019.

### Puzzle 5.1: Loading the Course Catalog Dataset from 2019

Using the name `df2019` instead of `df`, load the "Fall 2019 Course Catalog Dataset" that can be found at the following URL:
```
https://waf.cs.illinois.edu/discovery/course-catalog-fa19.csv
```

In [39]:
df2019 = pd.read_csv("/Users/scarlett_hjq/Downloads/course-catalog-fa19.csv")
df2019

Year  Term YearTerm Subject  Number  \
0     2019  Fall  2019-fa     AAS     100   
1     2019  Fall  2019-fa     AAS     105   
2     2019  Fall  2019-fa     AAS     120   
3     2019  Fall  2019-fa     AAS     199   
4     2019  Fall  2019-fa     AAS     200   
...    ...   ...      ...     ...     ...   
8584  2019  Fall  2019-fa    ZULU     202   
8585  2019  Fall  2019-fa    ZULU     403   
8586  2019  Fall  2019-fa    ZULU     404   
8587  2019  Fall  2019-fa    ZULU     405   
8588  2019  Fall  2019-fa    ZULU     406   

                                       Name  \
0              Intro Asian American Studies   
1     Introduction to Arab American Studies   
2             Intro to Asian Am Pop Culture   
3                Undergraduate Open Seminar   
4                      U.S. Race and Empire   
...                                     ...   
8584                     Elementary Zulu II   
8585                    Intermediate Zulu I   
8586                   Intermediate Zulu II   
8587                        Advanced Zulu I   
8588                       Advanced Zulu II   

                                            Description   Credit Hours  \
0     Interdisciplinary introduction to the basic co...       3 hours.   
1     Interdisciplinary introduction to the basic co...       3 hours.   
2     Introductory understanding of the way U.S. pop...       3 hours.   
3              May be repeated to a maximum of 6 hours.  1 TO 5 hours.   
4     Invites students to examine histories and narr...       3 hours.   
...                                                 ...            ...   
8584  Continuation of ZULU 201 with introduction of ...       5 hours.   
8585  Survey of more advanced grammar; emphasis on i...       4 hours.   
8586  Continuation of ZULU 403; emphasis on increasi...       4 hours.   
8587  Third year Zulu with emphasis on conversationa...       3 hours.   
8588  Continuation of Zulu 405 with increased emphas...       3 hours.   

                                           Section Info  \
0                                                   NaN   
1                                                   NaN   
2                                                   NaN   
3                                                   NaN   
4                                      Same as LLS 200.   
...                                                 ...   
8584  Same as AFST 252. Participation in the languag...   
8585                                                NaN   
8586                                                NaN   
8587                                                NaN   
8588                                                NaN   

                                      Degree Attributes  
0     Social & Beh Sci - Soc Sci, and Cultural Studi...  
1                Cultural Studies - US Minority course.  
2                Cultural Studies - US Minority course.  
3                                                   NaN  
4     Humanities - Hist & Phil, and Cultural Studies...  
...                                                 ...  
8584                                                NaN  
8585                                                NaN  
8586                                                NaN  
8587                                                NaN  
8588                                                NaN  

[8589 rows x 10 columns]

### Puzzle 5.2: Data Science Analysis

In the following cells,
- Select only the rows from the `IS` subject from the `df2019` dataset,
- Select only the rows from the `IS` subject from the `df` dataset (this contains the 2022 data),
- Complete the reflection question below (Puzzle 5.3).  Feel free to write more code to answer it.

In [40]:
df2019[df2019["Subject"] == "IS"]

Year  Term YearTerm Subject  Number  \
4843  2019  Fall  2019-fa      IS     107   
4844  2019  Fall  2019-fa      IS     199   
4845  2019  Fall  2019-fa      IS     202   
4846  2019  Fall  2019-fa      IS     310   
4847  2019  Fall  2019-fa      IS     351   
...    ...   ...      ...     ...     ...   
4930  2019  Fall  2019-fa      IS     591   
4931  2019  Fall  2019-fa      IS     592   
4932  2019  Fall  2019-fa      IS     593   
4933  2019  Fall  2019-fa      IS     594   
4934  2019  Fall  2019-fa      IS     599   

                                             Name  \
4843                       Data Science Discovery   
4844                   Undergraduate Open Seminar   
4845                     Social Aspects Info Tech   
4846                  Computing in the Humanities   
4847  The Design of Usable Information Interfaces   
...                                           ...   
4930                                    Practicum   
4931                            Independent Study   
4932                                  CAS Project   
4933                Information Sciences Practice   
4934                              Thesis Research   

                                            Description    Credit Hours  \
4843          Same as CS 107 and STAT 107. See STAT 107        4 hours.   
4844       Undergraduate Open Seminar. May be repeated.   1 TO 5 hours.   
4845       Same as INFO 202 and MACS 202. See INFO 202.        3 hours.   
4846  Explores use and application of technology to ...        3 hours.   
4847  Examines issues of Human Computer Interaction ...        3 hours.   
...                                                 ...             ...   
4930  Supervised field experience of professional-le...        2 hours.   
4931  Permits the intermediate or advanced student o...   2 TO 4 hours.   
4932  Individual study of a problem in library and i...   0 TO 8 hours.   
4933  Full-time or part-time practice of of any info...        0 hours.   
4934  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info  \
4843          Same as CS 107 and STAT 107. See STAT 107   
4844                                                NaN   
4845       Same as INFO 202 and MACS 202. See INFO 202.   
4846                                                NaN   
4847                  Prerequisite: Sophomore standing.   
...                                                 ...   
4930  2 graduate hours. No professional credit. Appr...   
4931  2 to 4 graduate hours. No professional credit....   
4932  0 to 8 graduate hours. No professional credit....   
4933  0 graduate hours. No professional credit. Appr...   
4934  0 to 16 graduate hours. No professional credit...   

                       Degree Attributes  
4843    Quantitative Reasoning I course.  
4844                                 NaN  
4845  Social & Beh Sci - Soc Sci course.  
4846                                 NaN  
4847                                 NaN  
...                                  ...  
4930                                 NaN  
4931                                 NaN  
4932                                 NaN  
4933                                 NaN  
4934                                 NaN  

[92 rows x 10 columns]

In [42]:
df[df["Subject"] == "IS"]

Year  Term YearTerm Subject  Number  \
6663  2022  Fall  2022-fa      IS     100   
6664  2022  Fall  2022-fa      IS     100   
6665  2022  Fall  2022-fa      IS     100   
6666  2022  Fall  2022-fa      IS     100   
6667  2022  Fall  2022-fa      IS     100   
...    ...   ...      ...     ...     ...   
6907  2022  Fall  2022-fa      IS     596   
6908  2022  Fall  2022-fa      IS     597   
6909  2022  Fall  2022-fa      IS     597   
6910  2022  Fall  2022-fa      IS     597   
6911  2022  Fall  2022-fa      IS     599   

                                                   Name  \
6663   Exploring the iSchool with a Human-Centered Lens   
6664   Exploring the iSchool with a Human-Centered Lens   
6665   Exploring the iSchool with a Human-Centered Lens   
6666   Exploring the iSchool with a Human-Centered Lens   
6667   Exploring the iSchool with a Human-Centered Lens   
...                                                 ...   
6907  Advanced Topics in Human-Centered Design &amp;...   
6908  Advanced Topics in Data Analytics &amp; Data S...   
6909  Advanced Topics in Data Analytics &amp; Data S...   
6910  Advanced Topics in Data Analytics &amp; Data S...   
6911                                    Thesis Research   

                                            Description    Credit Hours  \
6663  This course introduces students to the School ...        1 hours.   
6664  This course introduces students to the School ...        1 hours.   
6665  This course introduces students to the School ...        1 hours.   
6666  This course introduces students to the School ...        1 hours.   
6667  This course introduces students to the School ...        1 hours.   
...                                                 ...             ...   
6907  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6908  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6909  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6910  Variety of newly developed and advanced topics...   2 OR 4 hours.   
6911  Individual study and research. 0 to 16 graduat...  0 TO 16 hours.   

                                           Section Info Degree Attributes  \
6663  Prerequisite: Restricted to Majors Only; First...               NaN   
6664  Prerequisite: Restricted to Majors Only; First...               NaN   
6665  Prerequisite: Restricted to Majors Only; First...               NaN   
6666  Prerequisite: Restricted to Majors Only; First...               NaN   
6667  Prerequisite: Restricted to Majors Only; First...               NaN   
...                                                 ...               ...   
6907  2 or 4 graduate hours. No professional credit....               NaN   
6908  2 or 4 graduate hours. No professional credit....               NaN   
6909  2 or 4 graduate hours. No professional credit....               NaN   
6910  2 or 4 graduate hours. No professional credit....               NaN   
6911  0 to 16 graduate hours. No professional credit...               NaN   

      ... Section Status           Enrollment Status                Type  \
6663  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6664  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6665  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6666  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
6667  ...              A  CrossListOpen (Restricted)  Lecture-Discussion   
...   ...            ...                         ...                 ...   
6907  ...              A           Open (Restricted)              Online   
6908  ...              A           Open (Restricted)  Lecture-Discussion   
6909  ...              A           Open (Restricted)  Lecture-Discussion   
6910  ...              A           Open (Restricted)              Online   
6911  ...              A           Open (Restricted)   Independent Study   

     Type Code St

In [ ]:
# Feel free to add more code cells if needed by hovering over the area below the cell and pressing "+ Code"!

### Puzzle 5.3: Reflection on the Growth of Information Science

In the follow cell, write **at least one complete sentence explaining what you found** related to the growth of the IS program.  In your answer, **use actual data** (ex: by how much did it grow?)

IS only offered 92 rows of courses in 2019, but offers 249 rows of courses in 2022. We can find out that IS is in fast development and becomes more popular during these years.

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)